src/core/inductive.ml

```diff
@@ -0,0 +1,260 @@
+    [ind_sym] is applied and the variables of the products (in reverse order).
+    In case of a product, the function [get_name] splits the product and
+    naming the variable if it is not the case. In case of recursive occurren-
+    ces, the function [build_rec_hyp] builds the recursive hypothesis associa-
+    ted, and then the function [domrec] is applied to conclude the building.
+    Otherwise, the function [dom] is called. If you would like to store a tem-
+    porary result, the initial value is [acc], and you can change this value
+    in the recursive case with the function [f_rec], and on the other case
+    with the function [f_not_rec]. *)
+let fold_cons_typ :
+      popt ->
+      (term list -> 'b list -> 'a -> 'c) ->
+      ( (term, term) Bindlib.binder -> int -> 'b * term ) ->
+      (term list -> 'b -> 'a) ->
+      (term -> 'b -> 'a -> 'c -> 'c) ->
+      (term -> 'b -> 'c -> 'c) ->
+      sym -> sym ->
+      ('a -> 'b -> 'a -> 'a) ->
+      ('a -> 'b -> 'a) ->
+      'a ->
+      'c =
+  fun pos codom get_name build_rec_hyp domrec dom ind_sym
+      cons_sym f_rec f_not_rec acc->
+  let rec aux : 'b list -> 'a -> int -> term -> 'c = fun xs p i a ->
+    match Basics.get_args a with
+    | (Symb(s), ts) ->
+       if s == ind_sym then codom ts xs p
+       else fatal pos "%a is not a constructor of %a"
+              pp_symbol cons_sym pp_symbol ind_sym
+    | (Prod(a,b), _) ->
+       let (x,b) = get_name b i in
+       begin
+         match Basics.get_args a with
+         | (Symb(s), ts) ->
+             if s == ind_sym then
+               let rec_hyp = build_rec_hyp ts x in
+               let next = aux (x::xs) (f_rec p x rec_hyp) (i+1) b in
+               domrec a x rec_hyp next
+             else
+               let next = aux (x::xs) (f_not_rec p x) (i+1) b in
+               dom a x next
+         | _ -> fatal pos "The type of %a is not supported" pp_symbol cons_sym
+       end
+    | _ -> fatal pos "The type of %a is not supported" pp_symbol cons_sym
+  in aux [] acc 0 !(cons_sym.sym_type)
+
+(** [gen_rec_rules ind_sym rec_sym cons_list] returns the p_rules linking
+    with an induction principle [rec_sym] of the inductive type [ind_sym]
+    (with its constructors [cons_list]). *)
+let gen_rec_rules : sym -> sym -> sym list -> p_rule list =
+  fun ind_sym rec_sym cons_list ->
+
+  (* STEP 0: Define some tools which will be useful *)
+  let app : p_term -> p_term list -> p_term = List.fold_left P.appl in
+
+  (* STEP 1: Create the common head of the rules *)
+  let p_iden = P.iden rec_sym.sym_name in
+  let p_patt = P.patt0 "p" in
+  let head = P.appl p_iden p_patt in
+  let arg : sym list -> p_term = fun l ->
+    let rec aux : sym list -> p_term -> p_term = fun l acc ->
+      match l with
+      | []   -> acc
+      | t::q ->
+          let p_patt = P.patt0 ("pi" ^ t.sym_name) in
+          aux q (P.appl acc p_patt)
+    in
+    aux l head
+  in
+  let common_head = arg cons_list in
```

Comment:
This can be simplified as follows I believe:
```
let rec_iden = ... in
let p_patt = ... in
let add_arg t s = P.appl t (P.patt0 ("pi" ^ s.sym_name)) in
let head = P.appl rec_iden p_patt in
let common_head = List.fold_left add_arg head cons_list in
```

Reply:
Yes you right :) Thanks !